### PR TITLE
zerotierone: 1.8.4 -> 1.8.9

### PR DESCRIPTION
**Pull request by cmm**
###### Description of changes

ZerotierOne hasn't been refreshed in Nixpkgs for quite a while, and the updates since then promise some serious connection stability improvements.

Note that meanwhile ZerotierOne also grew SSO support written in Rust, so the change has to reflect that.  I also took this opportunity to get rid of `rec` in the derivation.

###### Things done

- Built on platform(s)
  - [v] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [v] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cmm** [comment]
re: the ofborg-eval error, is referencing `Cargo.lock` from the source verboten for some reason?  if so, I have no idea how to work around this.

**Comment by Mindavi** [comment]
builtins.fetchurl is indeed not allowed. But there should be alternatives (doesn't just doing fetchurl not work for a file).

**Comment by cmm** [comment]
> builtins.fetchurl is indeed not allowed. But there should be alternatives (doesn't just doing fetchurl not work for a file).

do elaborate?

**Comment by Mindavi** [comment]
Honestly, I'm not sure how to resolve this. I can confirm it builds locally though.

Closing since this won't be picked up for now. Hopefully someone else can look into this another time.

**Comment by cmm** [comment]
my current understanding is that this cannot work with hydra as it stands at all, since fetching a cargo.lock and deducing inputs from it is ifd, so it's actually nice of ofborg to fail the way it does.  welp.

File: pkgs/tools/networking/zerotierone/default.nix [hunk 1]
```diff
@@ -20,9 +42,22 @@ stdenv.mkDerivation rec {
       --replace 'armv5' 'armv6'
   '';
 
+  cargoDeps = rustPlatform.importCargoLock { inherit lockFile; };
```
Comment (misuzu) [inline pkgs/tools/networking/zerotierone/default.nix:45]: Please check if this works with ofborg:
```
  cargoDeps = rustPlatform.fetchCargoTarball {
    src = "${src}/zeroidc";
    name = "${pname}-${version}";
    sha256 = "sha256-PDsJtz279P2IpgiL0T92IbcANeGSUnGKhEH1dj9VtbM=";
  };
```

File: pkgs/tools/networking/zerotierone/default.nix [hunk 2]
```diff
@@ -20,9 +42,22 @@ stdenv.mkDerivation rec {
...
+  postPatch = "cp ${lockFile} Cargo.lock";
```
Comment (misuzu) [inline pkgs/tools/networking/zerotierone/default.nix:46]:
```
  postPatch = "cp ${src}/zeroidc/Cargo.lock Cargo.lock";
```

**Comment by cmm** [comment]
This pull is hereby resurrected as #173142.  Many thanks to @misuzu for his working suggestion!

File: pkgs/tools/networking/zerotierone/default.nix (hunk 1 above)
Comment (cmm) [reply pkgs/tools/networking/zerotierone/default.nix:45]: it does! :+1:
(and I'd like to understand why, academically)

**Pull request closed without merging**